### PR TITLE
[NPUW] Weights bank operate with UIDs

### PR DESCRIPTION
**Pull request by smirnov-alexey**

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 1]
```diff
@@ -35,38 +35,39 @@ class BankManager {
...
-    if (iter_device != device_bank.storage.end() && iter_device->second) {
-        // Already allocated
-        // tensor (the key) may be coming from a 2nd (3rd, ...) model
-        // detach it here just in case
-        const_cast<LazyTensor&>(tensor).detach();
-        return iter_device->second;
+    if (device_bank.registered_tensors.find(tensor) == device_bank.registered_tensors.end()) {
+        auto uid = uid_count++;
+        device_bank.registered_tensors[tensor] = uid;
+        device_bank.storage[uid] = {tensor, ov::Tensor()};
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:47]: Missing detach

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 2]
```diff
@@ -35,38 +35,39 @@ class BankManager {
...
     auto& device_bank = m_device_banks[device_for_alloc];
-    if (device_bank.storage.find(tensor) == device_bank.storage.end()) {
-        device_bank.storage[tensor] = ov::Tensor();
-    }
+
+    std::unique_lock<std::mutex> dev_guard(device_bank.mutex);
+    auto iter_device = device_bank.storage.find(uid);
+
+    NPUW_ASSERT(iter_device != device_bank.storage.end() && iter_device->second.tensor &&
+                "Tensor should be registered and allocated first!");
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:65]: I believe it should only be registerd at this point. Previously, if the device bank didn't have a device tensor, evaluation happened. Did this contract change now?

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 3]
```diff
@@ -106,7 +109,7 @@ ov::Tensor Bank::eval_and_alloc(const LazyTensor& tensor,
 
     std::unique_lock<std::mutex> guard(dbank.mutex);
     if (device_for_alloc == "CPU") {
-        dbank.storage[tensor] = transformed_tensor;
+        dbank.storage[dbank.registered_tensors[tensor]].tensor = transformed_tensor;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:112]:
```suggestion
        dbank.storage[dbank.registered_tensors.at(tensor)].tensor = transformed_tensor;
```

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 4]
```diff
@@ -117,7 +120,7 @@ ov::Tensor Bank::eval_and_alloc(const LazyTensor& tensor,
     remote_tensor =
         remote_ctx->create_host_tensor(transformed_tensor.get_element_type(), transformed_tensor.get_shape());
     allocated_tensor = ov::make_tensor(remote_tensor);
-    dbank.storage[tensor] = allocated_tensor;
+    dbank.storage[dbank.registered_tensors[tensor]].tensor = allocated_tensor;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:123]: here too

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.hpp [hunk 5]
```diff
@@ -26,18 +26,26 @@ class Bank {
...
     void evaluate_and_allocate();
-    bool is_remote(const LazyTensor& tensor) const;
+
+    bool is_remote(std::size_t uid) const;
 
 private:
+    struct StorageValue {
+        LazyTensor lt;
+        ov::Tensor tensor;
+    };
```
[the remark this answers is not in the document]
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.hpp]: Yep, same problem. Couldn't figure out a better way. Renamed as you suggested. Yes, we do need both lt and tensor (detach, allocate, etc)

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 2 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:66]: We had this discussion before during serialization review. I believe we agreed that we would drop the contract where `get()` could allocate anything unregistered

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.hpp [hunk 6]
```diff
@@ -26,18 +26,26 @@ class Bank {
...
-    // Allocate a new tensor (if needed) on a specified device. LazyTensor needs to be registered and evaluated first
-    ov::Tensor get(const LazyTensor& tensor, const std::string& device);
-    // Register LazyTensor in a bank if it's not there
-    void registerLT(const LazyTensor& tensor, const std::string& device);
+    // Register LazyTensor in a bank if it's not there. Returns LazyTensor's unique id
+    int64_t registerLT(const LazyTensor& tensor, const std::string& device);
+
+    // Allocate and evaluate a registered tensor on a specified device (if needed) and return it from the bank
+    ov::Tensor get(int64_t uid, const std::string& device);
+
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.hpp:34]: As you just mentioned, now it neither evaluates nor allocates the tensor. Just gets it

**Review by dmatveev — APPROVED** [review]
